src/Weather/Skysight/CDFDecoder.cpp

```diff
@@ -0,0 +1,334 @@
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "CDFDecoder.hpp"
+
+#ifdef ANDROID
+#include <netcdfcpp.h>
+#include <geotiffio.h>
+#include <xtiffio.h>
+#else
+#include <netcdf>
+#include <geotiff/geotiffio.h>
+#include <geotiff/xtiffio.h>
+#endif
+
+#include "SkysightAPI.hpp"
+#include "OS/FileUtil.hpp"
+#include "LogFile.hpp"
+
+void CDFDecoder::DecodeAsync()
+{
+  std::lock_guard<Mutex> lock(mutex);
+
+  Trigger();
+}
+
+void CDFDecoder::DecodeSync()
+{
+  Decode();
+}
+
+void CDFDecoder::Done()
+{
+  StandbyThread::LockStop();
+}
+
+CDFDecoder::Status CDFDecoder::GetStatus()
+{
+  std::lock_guard<Mutex> lock(mutex);
+  CDFDecoder::Status s = status;
+  return s;
+}
+
+void CDFDecoder::Tick() noexcept
+{
+  status = Status::Busy;
+  mutex.unlock();
+  Decode();
+}
+
+bool CDFDecoder::DecodeError() {
+  LogFormat("Decoder Error");
+  MakeCallback(false);
+  mutex.lock();
+  status = Status::Error;
+  return true;
+}
+
+bool CDFDecoder::DecodeSuccess()
+{
+  MakeCallback(true);
+  mutex.lock();
+  status = Status::Complete;
+  return true;
+}
+
+#ifdef ANDROID
+bool CDFDecoder::Decode()
+{
+  NcFile data_file(path.c_str(), NcFile::FileMode::ReadOnly);
+  if (!data_file.is_valid())
+    return DecodeError();
+  size_t lat_size = data_file.get_dim("lat")->size();
+  size_t lon_size = data_file.get_dim("lon")->size();
+
+  AllocatedArray<double>lat_vals(lat_size);
+  AllocatedArray<double>lon_vals(lon_size);
+  AllocatedArray<double>var_vals(lat_size * lon_size);
+
+  data_file.get_var("lat")->get(&lat_vals[0], lat_size);
+  data_file.get_var("lon")->get(&lon_vals[0], lon_size);
+
+  double lat_min = lat_vals[lat_size - 1];
+  double lat_max = lat_vals[0];
+  double lon_min = lon_vals[0];
+  double lon_max = lon_vals[lon_size - 1];
+  double lon_scale = (lon_max - lon_min) / lon_size;
+  double lat_scale = (lat_max - lat_min) / lat_size;
+
+  NcVar *data_var = data_file.get_var(data_varname.c_str());
+  if (!data_var->is_valid())
+    return DecodeError();
+
+  data_var->get(&var_vals[0], (long)lat_size, (long)lon_size);
+  double fill_value = data_var->get_att("_FillValue")->values()->as_double(0);
+  float var_offset = data_var->get_att("add_offset")->values()->as_float(0);
+  float var_scale = data_var->get_att("scale_factor")->values()->as_float(0);
+
+  //Generate GeoTiff
+  TIFF *tf = XTIFFOpen(output_path.c_str(), "w");
+  if (!tf)
+    return DecodeError();
```

Comment:
If an error occurs, throw an exception.

Reply:
LoadGeoTiff(Path path) throws an exception as below. Do same?
 
    auto gtif = GTIFNew(tiff.Get());
    if (gtif == nullptr)
      throw std::runtime_error("Not a GeoTIFF file");